components/l10n/common/locale_util.cc

```diff
@@ -5,78 +5,113 @@
+}
 
-}  // namespace
+inline std::string GetLocale(const icu::Locale& icu_locale) {
+  std::string locale{icu_locale.getBaseName()};
+  if (locale.empty()) {
+    return kDefaultLocale;
+  }
 
-std::string GetLanguageCode(
-    const std::string& locale) {
-  const std::vector<std::string> locale_components = base::SplitString(locale,
-      ".", base::KEEP_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
+  return locale;
+}
 
-  if (locale_components.empty()) {
+inline std::string GetLanguageCode(const icu::Locale& icu_locale) {
+  std::string language_code{icu_locale.getLanguage()};
+  if (language_code.empty()) {
     return kDefaultLanguageCode;
   }
 
-  std::string normalized_locale = locale_components.front();
-  std::replace(normalized_locale.begin(), normalized_locale.end(), '-', '_');
-
-  const std::vector<std::string> components = base::SplitString(
-      normalized_locale, "_", base::KEEP_WHITESPACE, base::SPLIT_WANT_ALL);
+  return language_code;
+}
 
-  if (components.empty()) {
-    return kDefaultLanguageCode;
+inline std::string GetCountryCode(const icu::Locale& icu_locale) {
+  std::string country_code{icu_locale.getCountry()};
+  if (country_code.empty()) {
+    return kDefaultCountryCode;
   }
 
-  const std::string language_code = components.front();
+  return country_code;
+}
 
-  return base::ToLowerASCII(language_code);
+absl::optional<std::string>& LocaleForTesting() {
+  static base::NoDestructor<absl::optional<std::string>> locale;
+  return *locale;
 }
 
-std::string GetCountryCode(
-    const std::string& locale) {
-  const std::vector<std::string> locale_components = base::SplitString(locale,
-      ".", base::KEEP_WHITESPACE, base::SPLIT_WANT_NONEMPTY);
+inline void SetDefaultICULocaleForTesting() {
+  CHECK(LocaleForTesting());
 
-  if (locale_components.empty()) {
-    return kDefaultCountryCode;
+  const icu::Locale icu_locale = GetNormalizedICULocale(*LocaleForTesting());
+  CHECK(!icu_locale.isBogus());
+
+  UErrorCode error_code = U_ZERO_ERROR;
+  icu::Locale::setDefault(icu_locale, error_code);
+  CHECK(U_SUCCESS(error_code));
+}
+
+// Workaround to set the default ICU locale across all processes for testing.
+inline void MaybeSetDefaultICULocaleForTesting() {
+  if (LocaleForTesting()) {
+    SetDefaultICULocaleForTesting();
   }
+}
 
-  std::string normalized_locale = locale_components.front();
-  std::replace(normalized_locale.begin(), normalized_locale.end(), '-', '_');
+}  // namespace
 
-  const std::vector<std::string> components = base::SplitString(
-      normalized_locale, "_", base::KEEP_WHITESPACE, base::SPLIT_WANT_ALL);
+void SetLocaleForTesting(const std::string& locale) {
+  LocaleForTesting() = locale;
+  SetDefaultICULocaleForTesting();
+}
 
-  if (components.size() <= 1) {
-    return kDefaultCountryCode;
+std::string GetLocale() {
+  MaybeSetDefaultICULocaleForTesting();
+  return GetLocale(icu::Locale::getDefault());
+}
+
+std::string GetLanguageCode() {
+  MaybeSetDefaultICULocaleForTesting();
```

Comment:
this code doesn't look good. you should not call `*ForTesting` methods from the production code.
is there some other way to do test-only setup?

Reply:
Fixed